packages/vite/src/node/optimizer/scan.ts

```diff
@@ -47,14 +47,92 @@ const htmlTypesRE = /\.(html|vue|svelte|astro|imba)$/
 export const importsRE =
   /(?<!\/\/.*)(?<=^|;|\*\/)\s*import(?!\s+type)(?:[\w*{}\n\r\t, ]+from)?\s*("[^"]+"|'[^']+')\s*(?=$|;|\/\/|\/\*)/gm
 
-export async function scanImports(config: ResolvedConfig): Promise<{
-  deps: Record<string, string>
-  missing: Record<string, string>
-}> {
+export function scanImports(config: ResolvedConfig): {
+  cancel: () => Promise<void>
+  result: Promise<{
+    deps: Record<string, string>
+    missing: Record<string, string>
+  }>
+} {
   // Only used to scan non-ssr code
 
   const start = performance.now()
+  const deps: Record<string, string> = {}
+  const missing: Record<string, string> = {}
+  let entries: string[]
+
+  const esbuildContext: Promise<BuildContext | undefined> = computeEntries(
+    config,
+  ).then((computedEntries) => {
+    entries = computedEntries
+
+    if (!entries.length) {
+      if (!config.optimizeDeps.entries && !config.optimizeDeps.include) {
+        config.logger.warn(
+          colors.yellow(
+            '(!) Could not auto-determine entry point from rollupOptions or html files ' +
+              'and there are no explicit optimizeDeps.include patterns. ' +
+              'Skipping dependency pre-bundling.',
+          ),
+        )
+      }
+      return
+    } else {
+      debug(`Crawling dependencies using entries:\n  ${entries.join('\n  ')}`)
+    }
+
+    return prepareEsbuildScanner(config, entries, deps, missing)
```

Comment:
could simplify this slightly. the `else` isn't needed since the `if` will `return`
```suggestion
    }

    debug(`Crawling dependencies using entries:\n  ${entries.join('\n  ')}`)
    return prepareEsbuildScanner(config, entries, deps, missing)
```